In [4]:
!pip install pysam

In [5]:
import pysam

bam_file = pysam.AlignmentFile("merged-tumor.bam", "rb")  

In [6]:
number_unmapped = 0
quality_zero = 0
total_quality = 0
total = 0

for read in bam_file:
  if read.is_unmapped:
    number_unmapped += 1

  if read.mapping_quality==0:
    quality_zero += 1
    
  total_quality += read.mapping_quality
  total += 1

  if total==1:
    first_read = read
    print("The first read from the AlignmentFile:")
    print(first_read, "\n")

avg_all = total_quality/total
avg_not_all = total_quality/(total-quality_zero)

print("Unmapped reads: ", number_unmapped)
print("Total number of reads: ", total)
print("Number of reads with mapping quality 0: ", quality_zero)
print("Average mapping quality for all the reads: ", avg_all)
print("Average mapping quality without reads with 0 mapp quality: ", avg_not_all, "\n")


The first read from the AlignmentFile:
C0HVYACXX120402:7:1207:5722:57044	1187	20	9483248	27	76M	20	9483381	76	TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG	array('B', [28, 28, 27, 29, 31, 30, 31, 31, 29, 31, 35, 30, 29, 31, 34, 30, 29, 23, 41, 32, 20, 30, 29, 34, 34, 29, 30, 31, 30, 30, 30, 33, 33, 26, 39, 12, 25, 19, 32, 30, 35, 28, 35, 33, 23, 33, 35, 36, 30, 38, 33, 41, 34, 35, 31, 33, 23, 30, 30, 36, 27, 32, 29, 34, 35, 41, 33, 31, 33, 29, 32, 32, 31, 31, 31, 34])	[('XA', 'GL000217.1,-110754,76M,1;'), ('BD', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('MD', '76'), ('RG', '1'), ('BI', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('NM', 0), ('MQ', 27), ('AS', 76), ('XS', 71)] 

Unmapped reads:  17765
Total number of reads:  2921629
Number of reads with mapping quality 0:  126628
Average mapping quality for all the reads:  55.91379158681681
Average mapping quality without reads wit

The first read from the BAM file represents an AlignedSegment and it containes the following fields:

*   QNAME 
*   FLAG
*   RNAME
*   POS
*   MAQ
*   CIGAR...

Specifically, the FLAG field respresents some additional information about the read. Here this field has value 1187 which is in binary representation 10010100011. The encoded properties of the read are:

* Read is paired
* Read is mapped in proper pair
* Mate reverse strand
* Second in pair
* Read is PCR or optical duplicate




